borg/cache.py

```diff
@@ -140,20 +142,39 @@ def create(self):
         with open(os.path.join(self.path, 'files'), 'wb') as fd:
             pass  # empty file
 
-    def _do_open(self):
-        self.config = configparser.ConfigParser(interpolation=None)
-        config_path = os.path.join(self.path, 'config')
-        self.config.read(config_path)
+    def _check_upgrade(self, config_path):
         try:
             cache_version = self.config.getint('cache', 'version')
-            wanted_version = 1
-            if cache_version != wanted_version:
-                self.close()
-                raise Exception('%s has unexpected cache version %d (wanted: %d).' % (
-                    config_path, cache_version, wanted_version))
+            if cache_version != CACHE_VERSION:
+                if cache_version == 1:
+                    # borg < 1.0.8rc1 (cache version 1) had different canonicalization,
+                    # upgrade it to current canonicalization (cache version 2):
+                    cache_loc = self.config.get('cache', 'previous_location', fallback=None)
+                    if cache_loc:
+                        repo_loc = self.repository._location.canonical_path()
+                        rl = Location(repo_loc)
+                        cl = Location(cache_loc)
+                        if cl.proto == rl.proto and cl.user == rl.user and cl.host == rl.host and cl.port == rl.port \
+                           and \
+                           cl.path and rl.path and \
+                           cl.path.startswith('/~/') and rl.path.startswith('/./') and cl.path[3:] == rl.path[3:]:
+                            # everything is same except the expected change in relative path canonicalization,
+                            # update previous_location to avoid warning / user query about changed location:
+                            self.config.set('cache', 'previous_location', repo_loc)
```

Comment:
hmm, thinking about it again, the flag / version change is not needed if we accept the false positive:

the longish if-clause uses the location from the current commandline (repository._location), so it can detect that change without having to check flag / version first.

If the commandline is ever changed from ssh://u@h/~/p to ssh://u@h/./p, this would result in a false positive and the "repo moved" warning would be suppressed.

I have to admit that I didn't investigate / understand yet why this check is done, I just remember that @jborg introduced this change together with the security fix for the exchange-to-unencrypted-empty-repo serverside attack.